src/netguard/tun.c

```diff
@@ -0,0 +1,181 @@
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/select.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <time.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+
+#define TIMEOUT_MS 300
+#define MIN_BUFFER_SIZE 17
+#define MAX_BUFFER_SIZE 213
+
+typedef enum {
+    SEND_RANDOM_DATA_OK,
+    SEND_RANDOM_DATA_ERROR_BIND_SOCKET,
+    SEND_RANDOM_DATA_ERROR_SEND_SOCKET
+} send_data_result_t;
+
+/**
+ * Returns whether the given address represents a public IP address or not
+ * @param address IPv4 or IPv6 address
+ * @return `true` if it's public IP address, false otherwise
+ */
+static bool is_public_ip(const struct sockaddr_storage* address);
+
+/**
+ * Sends random data to a random IP public IP address
+ * @param is_ipv6_enabled whether trying on IPv6 as well
+ * @return `send_data_result_t` that identifies syccess or error
+ */
+static send_data_result_t try_sending_random_udp(int is_ipv6_enabled);
+
+#define TIMEOUT_SECONDS 60
+#define TIMEOUT_MICROSECONDS 300000
+
+int wait_for_tunnel_up(int tun_fd, int is_ipv6_enabled) {
+    fd_set fd_set;
+    struct timeval timeout;
+    struct timeval start_time, current_time;
+
+    FD_ZERO(&fd_set);
+    FD_SET(tun_fd, &fd_set);
+
+    timeout.tv_sec = TIMEOUT_SECONDS;
+    timeout.tv_usec = TIMEOUT_MICROSECONDS;
+
+    gettimeofday(&start_time, NULL);
+
+    while (1) {
+        gettimeofday(&current_time, NULL);
+        long elapsed_seconds = current_time.tv_sec - start_time.tv_sec;
+        long elapsed_microseconds = current_time.tv_usec - start_time.tv_usec;
+
+        if (elapsed_seconds > TIMEOUT_SECONDS ||
+            (elapsed_seconds == TIMEOUT_SECONDS && elapsed_microseconds >= TIMEOUT_MICROSECONDS)) {
+            return -1; // Timeout occurred
+        }
+
+        int ready = select(tun_fd + 1, &fd_set, NULL, NULL, &timeout);
+        if (ready > 0) {
+            return 0; // Tunnel device is ready
+        }
+
+        FD_SET(tun_fd, &fd_set);
+        try_sending_random_udp(is_ipv6_enabled);
```

Comment:
Do we care if this succeeded or failed?

Reply:
Not really. This is just to ensure there's traffic to be routed through the tunnel. The important part is detecting if the `tun` is ready to read in the `select`